### ISSUE TITLE
Allow usage of the same attribute name in different branches within variant types

### ISSUE THREAD

**Issue by cryzed** [nim-lang/Nim#3629]
The following code snippet raises an error during compilation: `Error: redefinition of 'someAttribute'`. 

```nim
  type
    MyKind = enum
      Kind1, Kind2, Kind3

    MyType = ref object
      case kind: MyKind
      of Kind1:
        someAttribute: string
      of Kind2:
        someAttribute: string
      of Kind3:
        someOtherAttribute: int
```

Since only a single branch, depending on `kind`, can be active at once, it would seem logical for this to work -- since no actual redefinition of `someAttribute` ever takes place. Currently I need to work around this by naming the fields for each branch differently and creating a custom umbrella accessor proc that has to check the object kind at runtime and return the correct field, which is associated with its own problems: #3628.

**Comment by ephja** [comment]
Keep in mind that discriminators can be grouped together, though sometimes some fields need to be shared and others not.

**Comment by kirbyfan64** [comment]
What about:

```nim
of Kind1, Kind2:
    someAttribute: string
```

**Comment by cryzed** [comment]
And while we are at it, should it also be possible to have attributes of the same name but with a different type depending on the branch? This question was also raised in IRC and would improve [this code](https://github.com/nim-lang/Nim/blob/devel/lib/pure/json.nim#L557) for example.

**Comment by Araq** [comment]
> And while we are at it, should it also be possible to have attributes of the same name but with a different type depending on the branch?

No.

> This question was also raised in IRC and would improve this code for example, i.e. num and fnum.

No, it wouldn't.

**Comment by dom96** [comment]
> >   This question was also raised in IRC and would improve this code for example, i.e. num and fnum.

> No, it wouldn't.

IRC discussion [here](http://irclogs.nim-lang.org/08-12-2015.html) (Starts at 12:38:06)

**Comment by Araq** [comment]
Conclusion:

Alternatively we can weaken the disjoint condition for 'case' in 'object'.
So that this is allowed:

```nim
case kind: NodeKind
of foo, bar, baz: children: seq[Node]
of foo: additonalFieldForFoo: int
of bar: onlyvalidForBar: float
```

then it's much more obvious that 'children' is actually a *shared* field

**Comment by vegansk** [comment]
:+1: for this variant

**Comment by Varriount** [comment]
Wouldn't this cause divergence in the case statement syntax for type variants and conditions though?

**Comment by vegansk** [comment]
@Varriount, there is no divergence in Araq's variant. ``sizeof(x.children)`` and ``type(x.children)`` will always be the same

**Comment by Varriount** [comment]
@vegansk Ah, I see. However there is still the issue that we now have two different, subtle behaviors for case statements.

Enacting this change means that case blocks accept duplicate labels iff they are in a type definition, but not when they are in regular code. That's what I meant by divergent behavior.

**Comment by akbcode** [comment]
The limitation that two cases must either share all fields or none seems unreasonable.

**Comment by dom96** [comment]
> I'd much rather allow duplIcate member names in each of block, since that seems to be what users try first.

👍 to this

and 👎 to @Araq's conclusion

**Comment by cryzed** [comment]
I thought @Araq's reply was completely unproductive in the first place. IIRC he seemed more pissed off because he didn't want to consider implementing it. But I'll let this rest, it's nice that this is finally getting attention.

**Comment by Araq** [comment]
Dunno why you think that. My proposal would actually work, the other proposals are just nuts. I don't care how many thumbs down I get for it, you cannot vote for nonsense.

**Comment by GULPF** [comment]
@Araq What's the problem with the other proposal? If the fields with a shared name are required to also have the same type, then it accomplishes the same thing as your proposal, but with a more obvious syntax (at least to me).

**Comment by Bulat-Ziganshin** [comment]
The proposal in the first post is how things are done in Haskell, and it frequently became useful:

```
Entry = ref object
      case kind
      of File:
        name: string
        modified: datetime
        size: int64
      of Dir:
        name: string
        modified: datetime
      of Device:
        name: string
```

**Comment by Araq** [comment]
Under my proposal this would become:

```nim

Entry = ref object
      case kind
      of File, Dir, Device: name: string
      of File, Dir:
        modified: datetime
      of File:
        size: int64
```

which is clearer. IMO anyway.

**Comment by BarelyMalicious** [comment]
IMO the original proposal is better and more easily readable than @Araq 's proposal as over there I can easily visually parse which fields a particular branch has by just looking at that branch, whereas in Araq's proposal I have to look at the complete object to know that.

**Comment by Araq** [comment]
Yeah and in my proposal you can easily the "shared" aspect of the fields.

**Comment by dom96** [comment]
Less code duplication wins in my book. So I have to vote for @Araq's proposal. 

Especially since I assume implementing it is significantly easier.

**Comment by Bulat-Ziganshin** [comment]
Pro @Araq proposal:
- avoids redundancy
- easier to implement

Pro classic syntax:
- familiarity (it was used from Pascal to F#, while I don't know any language using Araq approach)
- when I know that current tag is File, I can quickly find which fields are available. I hardly ever had opposite task - i.e. figure out in which modes field "date" is available
- Currently there is no way to initialize object position-wise, i.e. File("io.c", 1, "1970-01-01"), rather than fieldname-wise, i.e. File(name:"io.c", size:1, date:"1970-01-01"). But if position-wise construction will be implemented, Araq approach will make it harder for human to find all fields for given tag and correctly use the constructor call.

So i feel that Araq proposal avoids redundancy (which is source of errors), but makes the type definition less readable. But all problems mentioned can be fixed with macroses. And while theoretically I can imagine various problems, practically it may be not an issue at all. The real definition in my archiver actually looks better with Araq style:

```
Entry = ref object
      case kind
      of File,Dir:
        name: string
        modified: datetime
        attr: uint32
        group: uint32
      of Dir:
        files: seq[Entry]
      of File:
        size: int64
        crc32: int32
```
So, my voice is to implement any proposal - this feature is important to have, while concrete syntax isn't of much issue. If users will be really unhappy with chosen syntax, alternative one may be added much later.

**Comment by andreaferretti** [comment]
I should mention that [patty](https://github.com/andreaferretti/patty#constructing-variant-objects) already supports a slightly more convenient syntax for defining variant types, and when this feature is available, it will be extended to support it, so it will translate

```nim
variant Entry:
  File(name: string, modified: datetime, size: int64)
  Dir(name: string, modified: datetime)
  Device(name: string)
```

into whatever syntax is chosen, for instance

```nim
type Entry = object
      case kind
      of File, Dir, Device: name: string
      of File, Dir:
        modified: datetime
      of File:
        size: int64
```

Whatever concrete syntax is chosen, one can always write a macro to improve on that, so I wouldn't worry too much about which syntax is best (both have advantages)

**Comment by Araq** [comment]
The problem with the non-Araq syntax is that it is totally misleading, in ML/Rust/etc the ``name`` field is always protected (can only be accessed by pattern matching as anything else would be unsafe) and **not** shared. You cannot access ``obj.name`` without a guard in ML while you can in Nim.

**Comment by Bulat-Ziganshin** [comment]
First, both proposals are semantically equivalent and can be mechanically translated to each other (as @andreaferretti plans to implement in patty). You can put restrictions on top of it according to your taste.

I worked only with Haskell, where

```Haskell
data Entry = File {name::String, size:: Int, attr:: Int}
           | Dir {name::String, attr:: Int, files:: [Entry]}
```
- defines constructor function `File "io.c" 0 0x0` and constructor with record syntax `File {name="io.c", size=0, attr=0x0}` (ditto for Dir)
- defines attribute access functions `name :: Entry -> String`, `files :: Entry -> [Entry]`... which fails when entry has improper tag
- allows to update attributes with record syntax `entry {files = []}` which fails when entry has improper tag, too
- allows pattern-matching with syntaxes similar to both constructors, i.e. `File aname asize _ -> print (aname,asize)` and `File {name=aname , size = asize} -> print (aname,asize)`

BTW, I just realized that people married to FP like me will anyway use patty and don't care about compiler syntax. And if classic syntax is really harder to deal, implementing @Araq syntax in compiler and classic syntax in library - will mean shifting one more complexity from compiler to library.

**Comment by jacereda** [comment]
How about this one?
```
type Entry = object
  kind: EntryKind

  case kind
  of File, Dir, Device:
    name: string

  case kind
  of File, Dir:
   modified: datetime

  case kind
  of File:
   size: int64
```

It's a bit cumbersome, but I suspect codegen could be easier that way. Things like the GC traversal will just generate as many `switch` statements as you have specified in your type.

**Comment by alehander42** [comment]
@Araq so is there an agreed design here? I remember you proposed using `if` instead of `case` for shared fields.
(I still think shared fields instead of kind=>fields is less useful & harder to codegen, but more important is to at least get one of those)

**Comment by alehander42** [comment]
@PMunch I don't think unfamilar syntax is the problem: one can use `if a.kind in {E, F}` and still make it: regular `case`(or this `case`) should not be changed to accomodate it IMO.

Deduplication might be nice, but on the other hand you'll easily forget to add all subbranches for the same value . 

Otherwise, it's just a matter of how often does one need to 

* see shared fields
* see fields for a given kind

I think the second usecase is way more useful and important: otherwise I need to always read the whole type to find out all possible fields for `A`.

However that's something that need measure
